File: lib/native/capture_events.dart
// lib/native/capture_events.dart
import '../utils/constants.dart';
import 'event_channel.dart';

/// All event type values sent over the capture-events channel.
enum CaptureEventType {
  screenshotTaken('captureEvents.screenshotTaken'),
  arSessionStart('captureEvents.arSessionStart'),
  arSessionEnd('captureEvents.arSessionEnd'),
  modelInteraction('captureEvents.modelInteraction');

  const CaptureEventType(this.wireValue);

  final String wireValue;

  static CaptureEventType? fromWireValue(String value) {
    for (final t in CaptureEventType.values) {
      if (t.wireValue == value) return t;
    }
    return null;
  }
}

/// Payload emitted for every capture event.
final class CaptureEventPayload {
  const CaptureEventPayload({
    required this.type,
    required this.timestamp,
    this.modelId,
    this.metadata,
  });

  factory CaptureEventPayload.fromMap(dynamic raw) {
    final map = raw as Map<dynamic, dynamic>;
    final wireType = map['type'] as String? ?? '';
    return CaptureEventPayload(
      type: CaptureEventType.fromWireValue(wireType) ??
          CaptureEventType.screenshotTaken,
      timestamp: map['timestamp'] as int,
      modelId: map['modelId'] as String?,
      metadata: (map['metadata'] as Map<dynamic, dynamic>?)
          ?.cast<String, Object?>(),
    );
  }

  final CaptureEventType type;

  /// Epoch milliseconds when the event occurred.
  final int timestamp;

  /// ID of the model involved, if applicable.
  final String? modelId;

  /// Arbitrary additional event metadata.
  final Map<String, Object?>? metadata;
}

/// Typed [NativeEventChannel] for AR/capture lifecycle events.
class CaptureEventsChannel extends NativeEventChannel<CaptureEventPayload> {
  CaptureEventsChannel()
      : super(AppConfig.channelCaptureEvents, CaptureEventPayload.fromMap);
}

File: maya-viewer/src/native/captureEvents.ts
// src/native/captureEvents.ts
import type { EventChannel } from './eventChannel';

/** All capture event type constants. Keys used internally; values are the wire strings. */
export const CAPTURE_EVENT_TYPES = {
  SCREENSHOT_TAKEN: 'captureEvents.screenshotTaken',
  AR_SESSION_START: 'captureEvents.arSessionStart',
  AR_SESSION_END: 'captureEvents.arSessionEnd',
  MODEL_INTERACTION: 'captureEvents.modelInteraction',
} as const;

/** Union of all possible capture event type strings. */
export type CaptureEventType =
  (typeof CAPTURE_EVENT_TYPES)[keyof typeof CAPTURE_EVENT_TYPES];

/** Payload emitted for every capture event. */
export interface CaptureEventPayload {
  type: CaptureEventType;
  /** Epoch milliseconds at the time the event occurred. */
  timestamp: number;
  /** ID of the model involved, if applicable. */
  modelId?: string;
  /** Arbitrary additional event metadata. */
  metadata?: Record<string, unknown>;
}

/** Typed alias for a capture-events channel. */
export type CaptureEventsChannel = EventChannel<CaptureEventPayload>;
